daemon/snapper/snapper.go

```diff
@@ -257,21 +278,62 @@ func snapshot(a args, u updater) state {
 			progress.state = SnapStarted
 		})
 
-		l.Debug("create snapshot")
-		err := zfs.ZFSSnapshot(fs, snapname, false)
-		if err != nil {
-			hadErr = true
-			l.WithError(err).Error("cannot create snapshot")
+		var doneAt time.Time
+
+		var snapErr error
+		var preHookErr []error
+		var postHookErr []error
+		hookEnvExtra := map[string]string{
+			"fs":       fs.ToString(),
+			"snapname": snapname,
+		}
+
+		l.Debug("pre-snapshot hook")
+		preHookErr, hadErr = hookRun.RunFilesystemHooks(fs, hooks.Pre, hookEnvExtra)
+		for _, err := range preHookErr {
+			l.WithError(err).Error("cannot run pre-snapshot hook(s)")
+		}
+
+		if !hadErr {
+			l.Debug("create snapshot")
+			snapErr = zfs.ZFSSnapshot(fs, snapname, false)
+			if snapErr != nil {
+				hadErr = true
+				l.WithError(snapErr).Error("cannot create snapshot")
+			}
+			doneAt = time.Now()
+
+			if !hadErr {
+				l.Debug("post-snapshot hook")
+				var postHookHadFatalErr bool
+				postHookErr, postHookHadFatalErr = hookRun.RunFilesystemHooks(fs, hooks.Post, hookEnvExtra)
+				for _, err := range postHookErr {
+					l.WithError(err).Error("cannot run post-snapshot hook(s)")
+					if postHookHadFatalErr {
+						l.Error("the post-hook run was aborted due to a fatal error in the previous hook")
```

Comment:
`hadErr` really could be renamed to `hadSnapErr`, because all the following logic takes actions that relate to retrying a snapshot (i.e. state `ErrorWait`). I did not set `hadErr` here, because it seemed that the only consequences of a failed post-hook with `ErrIsFatal == true` should be to skip following post-hooks and log the error.

Reply:
> it seemed that the only consequences of a failed post-hook with ErrIsFatal == true should be to skip following post-hooks and log the error.

See the review-wide commend on Pre-only Pre-and-Post, and Post-Only hooks.